### ISSUE TITLE
pkg/dartdev/test/commands/run_test flaky -> RuntimeError

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#47700]
There are new test failures on [[dds] Add DAP support for 'hotReload' custom request that calls reloadSources](https://github.com/dart-lang/sdk/commit/51df1855ff7cc7959d76a57ebacf0a99bf930700).

The tests
```
pkg/dartdev/test/commands/run_test RuntimeError (expected Pass)
```
are failing on configurations
```
unittest-asserts-release-win
```

/cc @DanTup

**Comment by DanTup** [comment]
@bkonyi it looks like these failures are in `run_test.dart`:

https://dart-ci.firebaseapp.com/#showLatestFailures=true&configurations=unittest-asserts-release-win

I don't think any code in the CL above should be running for those tests at all - could these failures be unrelated flakes?

**Comment by bkonyi** [comment]
It looks like this is still an issue, even after https://github.com/dart-lang/sdk/commit/2408d388782826227a32df5e59a769101819edca. @athomas @whesse, do either of you have any other suggestions on how to resolve this?

I'm going to mark this test as approved again in the meantime.

**Comment by whesse** [comment]
I think the test could catch the exception thrown during the deletion of the temp directory, and ignore it.  Failing to delete a temp directory created for a test is not a failure of the test, and should not be a fatal infrastructure failure, I think.

On our CI bots, the directories might be kept by a local cache of that builder directory, which is reused, but the builders get recycled within a day on linux and on windows.  It would be nice for every test to always successfully clean up its working directory, but I think we can allow an occasional failure to do it, which is not registered as a test failure or an infra failure.

**Comment by athomas** [comment]
I wasn't clear enough in my first comment in the gardening doc, reposting the last comment on that thread:
> I think you need `Process.start`, because with `run` you also don't get the pid to kill in the teardown. If those processes start other processes there might be a need to track those and kill 'em too (one way would be by using OS specific ways to kill process trees).

Just using `run` instead of `runSync` doesn't make the test stop leaking processes. Leaving the temp directories around will turn the build purple (because there will still be a process preventing the deletion when the infra tries to clean up temp directories).

The solution is to use `Process.start`, track all started processes and kill them in the tear down (or when the test timed out). The test is leaking processes, that's the real problem.

**Comment by bkonyi** [comment]
> I wasn't clear enough in my first comment in the gardening doc, reposting the last comment on that thread:
> 
> > I think you need `Process.start`, because with `run` you also don't get the pid to kill in the teardown. If those processes start other processes there might be a need to track those and kill 'em too (one way would be by using OS specific ways to kill process trees).
> 
> Just using `run` instead of `runSync` doesn't make the test stop leaking processes. Leaving the temp directories around will turn the build purple (because there will still be a process preventing the deletion when the infra tries to clean up temp directories).
> 
> The solution is to use `Process.start`, track all started processes and kill them in the tear down (or when the test timed out). The test is leaking processes, that's the real problem.

[I'm using `Process.start` and killing the process on tear down](https://github.com/dart-lang/sdk/commit/2408d388782826227a32df5e59a769101819edca#diff-47d9dad19d88b6267fe4a76145e1d2e5341c1920a1166f84d7cf89dac4871240R108) in 2408d38 :-) However, that still doesn't seem to fix this issue, hence the [followup CL](https://dart-review.googlesource.com/c/sdk/+/222386) to catch the exception on Windows.

**Comment by athomas** [comment]
@bkonyi Sorry, I should've looked closer... 🤦 

I wonder if the problem is in `dispose()`:
```
  void dispose() {
    _process?.kill(); // <-- doesn't block until the process is dead.
    _process = null;
    if (dir.existsSync()) {
      dir.deleteSync(recursive: true); // <-- process might still be running and locking the directory.
    }
  }
```

So we probably need to await the exit code before deleting the directory.

**Comment by alexmarkov** [comment]
This test starts failing on `pkg-win-release`:

```
01:12 [32m+17[0m[33m ~1[0m[31m -1[0m: run DDS explicit enable dart run simple [1m[31m[E][0m[0m

  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\a4ba6124f' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7                 _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7             FileSystemEntity.deleteSync
  pkg\dartdev\test\utils.dart 100:11                TestProject.dispose
  pkg\dartdev\test\commands\run_test.dart 26:21     run.<fn>
  dart:async/future.dart 296:31                     new Future.sync
  package:test_api/src/backend/invoker.dart 236:18  Invoker.runTearDowns.<fn>.<fn>
  package:test_api/src/backend/invoker.dart 257:15  Invoker._waitForOutstandingCallbacks.<fn>
  package:test_api/src/backend/invoker.dart 254:14  Invoker._waitForOutstandingCallbacks.<fn>
  dart:async/zone.dart 1426:13                      _rootRun
  dart:async/zone.dart 1328:19                      _CustomZone.run
  dart:async/zone.dart 1861:10                      _runZoned
  dart:async/zone.dart 1783:10                      runZoned
  package:test_api/src/backend/invoker.dart 254:5   Invoker._waitForOutstandingCallbacks
  package:test_api/src/backend/invoker.dart 235:9   Invoker.runTearDowns.<fn>
  package:test_api/src/backend/invoker.dart 230:21  Invoker.runTearDowns.<fn>
  dart:async/zone.dart 1426:13                      _rootRun
  dart:async/zone.dart 1328:19                      _CustomZone.run
  dart:async/zone.dart 1861:10                      _runZoned
  dart:async/zone.dart 1783:10                      runZoned
  package:test_api/src/backend/invoker.dart 230:12  Invoker.runTearDowns
  package:test_api/src/backend/invoker.dart 398:52  Invoker._onRun.<fn>.<fn>.<fn>.<fn>
  package:test_api/src/backend/invoker.dart 257:15  Invoker._waitForOutstandingCallbacks.<fn>
  package:test_api/src/backend/invoker.dart 254:14  Invoker._waitForOutstandingCallbacks.<fn>
  dart:async/zone.dart 1426:13                      _rootRun
  dart:async/zone.dart 1328:19                      _CustomZone.run
  dart:async/zone.dart 1861:10                      _runZoned
  dart:async/zone.dart 1783:10                      runZoned
  package:test_api/src/backend/invoker.dart 254:5   Invoker._waitForOutstandingCallbacks
  package:test_api/src/backend/invoker.dart 398:17  Invoker._onRun.<fn>.<fn>.<fn>
  

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
```

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/18913/pkg/dartdev/test/commands/run_test

**Comment by rakudrama** [comment]
Started failing again.
```
01:24 [32m+20[0m[33m ~1[0m[31m -1[0m: run DevTools dart simple [1m[31m[E][0m[0m

  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\af619c7b7' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7              _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7          FileSystemEntity.deleteSync
  pkg\dartdev\test\utils.dart 101:11             TestProject.dispose
  ===== asynchronous gap ===========================
  pkg\dartdev\test\commands\run_test.dart 26:24  run.<fn>
```
https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/18983/pkg/dartdev/test/commands/run_test

I don't see how my change could cause this error.

**Comment by alexmarkov** [comment]
Failed once again:

```
01:40 [32m+21[0m[33m ~1[0m[31m -3[0m: run DevTools dart run disabled [1m[31m[E][0m[0m

  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\acf7602be' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7              _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7          FileSystemEntity.deleteSync
  pkg\dartdev\test\utils.dart 101:11             TestProject.dispose
  ===== asynchronous gap ===========================
  pkg\dartdev\test\commands\run_test.dart 26:24  run.<fn>
  ```

https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8827020818389088193/+/u/test_results/new_test_failures__logs_

**Comment by aam** [comment]
and again https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19018/pkg/dartdev/test/commands/run_test

**Comment by bkonyi** [comment]
@athomas @whesse, it looks like we're still having issues here, even after waiting on the process exit code. Should we maybe consider catching the exception since this isn't a symptom of an actual test failure?

**Comment by aam** [comment]
and again on https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8825685537532347809/+/u/test_results/new_test_failures__logs_

**Comment by aam** [comment]
Got https://dart-review.googlesource.com/c/sdk/+/226980 with the "fix". I am able to reproduce issue locally and confirm that cl addresses the problem.

**Comment by whesse** [comment]
The fix made another test change from Pass to flaky when it landed, and now to consistently failing with RuntimeError:
dartdev/test/analysis_server_test on Windows.

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19284/pkg/dartdev/test/analysis_server_test

**Comment by aam** [comment]
Thanks @whesse . I think that should be logged as a new separate issue.
